File: app/Domain/Identity/UserIdentity.php
<?php

namespace App\Domain\Identity;

use App\Domain\DecisionProjectionBase;
use App\Domain\IEventPublisher;

class UserIdentity {
    /** @var DecisionProjection */
    private $decisionProjection;

    public static function register(IEventPublisher $eventPublisher, $userId) {
        $eventPublisher->publish(new UserRegistered($userId));
    }

    /**
     * @param array $events
     */
    public function __construct($events) {
        $this->decisionProjection = new DecisionProjection($events);
    }

    public function logIn(IEventPublisher $eventPublisher) {
        $eventPublisher->publish(new UserConnected($this->decisionProjection->getUserId(), uniqid(), new \DateTime()));
    }
}

class DecisionProjection extends DecisionProjectionBase {

    /** @var string */
    private $userId;

    /**
     * @param array $events
     */
    public function __construct($events) {
        $this->registerUserRegistered();
        parent::__construct($events);
    }

    /**
     * @return string
     */
    public function getUserId()
    {
        return $this->userId;
    }

    private function registerUserRegistered()
    {
        $this->register('App\Domain\Identity\UserRegistered', function(UserRegistered $event) {
            $this->userId = $event->getUserId();
        });
    }
}

File: src/main/scala/domain/identity/UserIdentity.scala
package domain.identity

import java.time.{LocalDateTime, ZoneOffset}

import domain.EventPublisher
import domain.identity.event.{UserConnected, UserRegistered}

case class UserIdentity(userRegistered: UserRegistered) {
  import UserIdentity._

  private val projection=DecisionProjection.of(userRegistered)

  def logIn()(implicit ep:EventPublisher):Unit= {
    val sessionId = SessionId()
    ep.publish(UserConnected(sessionId, projection.userId, LocalDateTime.now(ZoneOffset.UTC)))
  }
}

object UserIdentity {
  def register(userId: UserId)(implicit ep:EventPublisher)=
    ep.publish(UserRegistered(userId))

  private case class DecisionProjection(userId: UserId)

  private object DecisionProjection{
    def of(userRegistered: UserRegistered)=DecisionProjection(userRegistered.userId)
  }
}
